pkg/controller/networkpolicy/tier.go

```diff
@@ -0,0 +1,145 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package networkpolicy provides NetworkPolicyController implementation to manage
+// and synchronize the Pods and Namespaces affected by Network Policies and enforce
+// their rules.
+
+package networkpolicy
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+
+	secv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/security/v1alpha1"
+)
+
+var (
+	// maxSupportedTiers is the soft limit on the maximum number of supported
+	// Tiers.
+	maxSupportedTiers = 50
+	// defaultTierPriority maintains the lowest priority for the system generated
+	// default Tier.
+	defaultTierPriority = int32(250)
+	// priorityMap maintains the Tier priority associated with system generated
+	// Tier names.
+	priorityMap = map[string]int32{
+		"application": defaultTierPriority,
+		"platform":    int32(150),
+		"networkops":  int32(100),
+		"securityops": int32(50),
+		"emergency":   int32(5),
+	}
+	// staticTierSet maintains the names of the static tiers such that they can
+	// be converted to corresponding Tier CRD names.
+	staticTierSet = sets.NewString("Emergency", "SecurityOps", "NetworkOps", "Platform", "Application")
+	// systemGeneratedTiers are the Tier CRs to be created at init.
+	systemGeneratedTiers = []*secv1alpha1.Tier{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "application",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["application"],
+				Description: "[READ-ONLY]: System generated default Application Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "platform",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["platform"],
+				Description: "[READ-ONLY]: System generated Platform Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "networkops",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["networkops"],
+				Description: "[READ-ONLY]: System generated NetworkOps Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "securityops",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["securityops"],
+				Description: "[READ-ONLY]: System generated SecurityOps Tier",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "emergency",
+			},
+			Spec: secv1alpha1.TierSpec{
+				Priority:    priorityMap["emergency"],
+				Description: "[READ-ONLY]: System generated Emergency Tier",
+			},
+		},
+	}
+)
+
+// InitializeTiers initializes the default Tiers created by Antrea on init. It
+// will first attempt to retrieve the Tier by it's name from K8s and if missing,
+// create the CR. InitializeTiers will be called as part of a Post-Start hook
+// of antrea-controller's APIServer.
+func (n *NetworkPolicyController) InitializeTiers() {
```

Comment:
I remember you said the default tiers can be renamed too? But here we just check tiers by name and recreate them if not found?

Reply:
Ok. I should ask could they be deleted? I remember you said yes.

Reply:
oh yes.. for that after the meeting i realized every time controller is restarted .. it will attempt to recreate the deleted resources because there is no way to understand whether a Tier was intentionally deleted by the user or it was missing on start up.. so to keep the behavior of system generated tiers uniform i made them read-only

Reply:
Ok. Let us keep this behavior then.